modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/utils/odata.es.js

```diff
@@ -0,0 +1,345 @@
+ * @param {string} queryString
+ * @returns {object} Criteria representation of the query string
+ */
+function translateQueryToCriteria(queryString) {
+	let criteria;
+
+	try {
+		const oDataASTNode = oDataFilterFn(queryString);
+
+		const criteriaArray = toCriteria({oDataASTNode});
+
+		criteria = isCriteriaGroup(criteriaArray[0]) ?
+			criteriaArray[0] :
+			wrapInCriteriaGroup(criteriaArray);
+	}
+	catch (e) {
+		criteria = null;
+	}
+
+	return criteria;
+}
+
+/**
+ * Recursively transforms the AST generated by the odata-v4-parser library into
+ * a shape the criteria builder expects. Returns an array so that left and right
+ * arguments can be concatenated together.
+ * @param {object} context
+ * @param {object} context.oDataASTNode
+ * @returns Criterion representation of an AST expression node in an array
+ */
+function toCriteria(context) {
+	const {oDataASTNode} = context;
+
+	const expressionName = getExpressionName(oDataASTNode);
+
+	let criterion;
+
+	if (isValueType(RELATIONAL_OPERATORS, expressionName)) {
+		criterion = transformOperatorNode(context);
+	}
+	else if (isValueType(CONJUNCTIONS, expressionName)) {
+		criterion = transformConjunctionNode(context);
+	}
+	else if (expressionName === GROUP) {
+		criterion = transformGroupNode(context);
+	}
+
+	return criterion;
+}
+
+/**
+ * Transforms conjunction expression node into a criterion for the criteria
+ * builder. If it comes across a grouping sharing an AND and OR conjunction, it
+ * will add a new grouping so the criteria builder doesn't require a user to
+ * know operator precedence.
+ * @param {object} context
+ * @param {object} context.oDataASTNode
+ * @returns an array containing the concatenated left and right values of a
+ * conjunction expression or a new grouping.
+ */
+function transformConjunctionNode(context) {
+	const {oDataASTNode} = context;
+
+	const conjunctionType = oDataASTNode.type;
+	const nextNode = oDataASTNode.value;
+
+	return hasDifferentConjunctions(context) ?
+		toCriteria(addNewGroup(context)) :
+		[...toCriteria(
+			{
+				oDataASTNode: nextNode.left,
+				prevConjunction: conjunctionType
+			}
+		),
+		...toCriteria(
+			{
+				oDataASTNode: nextNode.right,
+				prevConjunction: conjunctionType
+			}
+		)];
+}
+
+/**
+ * Transforms a group expression node into a criterion for the criteria
+ * builder. If it comes across a grouping that is redundant (doesn't provide
+ * readability improvements, superfluous to order of operations), it will remove
+ * it.
+ * @param {object} context
+ * @param {object} context.oDataASTNode
+ * @param {string} context.prevConjunction
+ * @returns Criterion representation of an AST expression node in an array
+ */
+function transformGroupNode(context) {
+	const {oDataASTNode, prevConjunction} = context;
+
+	return isRedundantGroup(context) ?
+		toCriteria(skipGroup(context)) :
+		[{
+			conjunctionName: getConjunctionForGroup(oDataASTNode),
+			groupId: generateGroupId(),
+			items: [...toCriteria(
```

Comment:
Do you need both the `[]` and `...`?  Unless `toCriteria` doesn't return a new array every time.

Reply:
`toCriteria` should return an array so I'll remove the `[]` and `...` 👍